### ISSUE TITLE
Support external $ref in JSON Schema

### ISSUE THREAD

**Issue by TyOverby** [quicktype/quicktype#522]
JSON Schema file in question: http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json

Running `quicktype -s schema deploymentTemplate.json -o deps.ts` yields a rather small [output file](https://gist.github.com/TyOverby/465126532210ed72f9b906c4f7e7e234).  

You'll notice the `resources: any[]` field.  This is where the bulk of the schema code should be.

Fair warning: once the URL links fully resolved, this file is massive. I wrote a compiler for it once, and it produced tens of thousands of lines of typescript.

**Comment by schani** [comment]
We don't support external `$ref` references and `allOf` yet, unfortunately.

What types would you expect this file to generate?

**Comment by TyOverby** [comment]
The "expected" output is too big to reason about, but I would expect that the typescript would successfully typecheck all the json that the schema typechecks.

**Comment by schani** [comment]
I'm curious: if you only care about validation, why wouldn't you use a schema validator like [ajv](https://github.com/epoberezkin/ajv)?

**Comment by TyOverby** [comment]
Oh, it's because I'd like the really nice tooling in Typescript that you get from having the schema translated.

I should add that I'm not really worried about this bug, like I said, I'm currently using a custom json-schema->typescript compiler that I made for this file specifically.  I'd like to move off it, and quicktype looks neat, but no rush.

**Comment by TyOverby** [comment]
Here's the same file, but with all of the $ref references inlined.  Warning, it's 93,000 lines.

 https://gist.githubusercontent.com/TyOverby/2f4a7036a511684c99accba28128e204/raw/29f25d3067dc7b28a60030e5b256813ea925f5d3/deploymentTemplate.json

**Comment by schani** [comment]
Thanks!

How do you compile `allOf`?

**Comment by TyOverby** [comment]
`allOf: [ a, b, c ]` becomes `a & b & c`

**Comment by schani** [comment]
@TyOverby I implemented `all-of` in #574. It doesn't completely accept your schema yet, because the schema contains enums with integers and booleans, which I haven't implemented yet.  I removed all those [here](https://gist.github.com/schani/47d1fbccb753cc403d6efd3943e26d77).  The types I get look a LOT simpler than the schema, so I'm very skeptical.  Could you please provide me with a couple of JSON files that satisfy that schema or, if that's not possible, test it yourself and give me feedback?

**Comment by TyOverby** [comment]
@schani Yeah, here's a repo filled with them! 

https://github.com/Azure/azure-quickstart-templates

Every one of those folders has an `azuredeploy.json` file that (should) satisfy that schema.

**Comment by schani** [comment]
@TyOverby I went over all those JSONs with [ajv](https://github.com/epoberezkin/ajv), and most of them do not satisfy the schema.  [Here's a list](https://gist.github.com/schani/d78ff39314425b4e31009b4114ef63e4) of the ones that do.

**Comment by schani** [comment]
@TyOverby Here's one oddity in your schema: There are quite a few places where you specify a `type` of `boolean`, but then for the same type specify an `enum` containing strings.  Example:

```json
{
    "type": "boolean",
    "enum": [
        "true"
    ]
}
```

This doesn't match any value, which implies that if a value is present in the JSON for that type, the JSON is invalid with respect to the schema.

I'm right now working on making your schema work with quicktype - all the pieces are finally in place - but we won't support "invalid" types for the time being.  I'm fixing it by changing the strings to the booleans I'm assuming they should represent.

**Comment by schani** [comment]
You're doing the same thing with `integer`s.

**Comment by TyOverby** [comment]
Yeah, honestly this isn't super surprising.  This massive schema is built from multiple schamas authored by dozens of teams across Microsoft Azure.  

Do you have an easy way of getting a list of the paths for these non-conformant types?  I could file bugs internally.

**Comment by schani** [comment]
@TyOverby No easy way right now.  If you just search for `"true"`, `"false"` and `"0"` I think should get them all.

**Comment by schani** [comment]
@TyOverby My latest working branch supports most of the valid JSONs now, with the caveat that we end up with a little too many `any` types than I'd like - I need to figure out a way to debug where in the schema they come from.  The JSONs that don't work are these:

```
101-redis-cache-azuredeploy.json
101-webappazure-oms-monitoring-azuredeploy.json
201-redis-premium-cluster-diagnostics-azuredeploy.json
201-redis-premium-persistence-azuredeploy.json
201-redis-premium-vnet-cluster-diagnostics-azuredeploy.json
```

The problem in all of them is the same: A resource with `type` of either `Microsoft.Cache/redis/providers/diagnosticsettings` or `runbooks`, both of which I couldn't find in the Schema, so I figured quicktype might be onto something.  Could you look into that, please?

I'll also have to re-run the validation with the Schema with the enums fixed.

**Comment by TyOverby** [comment]
@dpwrussell: We had success with the translation into typescript for the specific parts of the azure schema that we cared about.  You can find an old version of the code here: https://github.com/TyOverby/arm-compiler

Unfortunately we weren't able to use quicktype at the time, so I wrote my own compiler, which you can find pieces of in the project above.

**Comment by schani** [comment]
I retested this on current master (commit 662ad5b6) with `quicktype -s schema deploymentTemplate.json -o deps.ts`, and the original problem described here is gone. Back in 2018 quicktype didn't support external `$ref` or `allOf`, so it silently fell back to `resources: any[]` and produced a tiny, useless output. Both of those were implemented since (allOf landed in #574, plus external-ref resolution), so quicktype now actually follows the 300+ external refs in the Azure deployment template schema and generates real types for the `resources` structures instead of collapsing them to `any[]`. It's a heavy, network-bound run given the size of the schema, but it no longer falls short in the way originally reported here.

This will be in the next release. If you (or anyone hitting this later) can still reproduce the `any[]` collapse on a current build, please reopen with the schema/version you used and I'll take another look.
